src/lib/y2storage/planned/can_be_formatted.rb

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env ruby
+#
+# encoding: utf-8
+
+# Copyright (c) [2015-2017] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+
+module Y2Storage
+  module Planned
+    # Mixing for planned devices that can have an associated block filesystem.
+    # @see Planned::Device
+    module CanBeFormatted
+      # @return [Filesystems::Type] the type of filesystem this device should
+      #   get, like Filesystems::Type::BTRFS or Filesystems::Type::SWAP. A value of
+      #   nil means the device will not be formatted.
+      attr_accessor :filesystem_type
+
+      # @return [String] label to enforce in the filesystem
+      attr_accessor :label
+
+      # @return [String] UUID to enforce in the filesystem
+      attr_accessor :uuid
+
+      # @return [Array<Planned::BtrfsSubvolume>] Btrfs subvolumes
+      attr_accessor :subvolumes
+
+      # @return [String] Parent for all Btrfs subvolumes (typically "@")
+      attr_accessor :default_subvolume
+
+      # Initializations of the mixin, to be called from the class constructor.
+      def initialize_can_be_formatted
+      end
+
+      # Creates a filesystem for the planned device on the specified real
+      # BlkDevice object.
+      #
+      # This also sets all the filesystem attributes, like the mount point, and
+      # creates the corresponding Btrfs subvolumes if needed.
+      #
+      # Do nothing if #filesystem_type is not set.
+      #
+      # @param blk_dev [BlkDevice]
+      #
+      # @return [Filesystems::BlkFilesystem] filesystem
+      def create_filesystem(blk_dev)
+        return nil unless filesystem_type
+
+        filesystem = blk_dev.create_blk_filesystem(filesystem_type)
+        filesystem.mountpoint = mount_point if mount_point && !mount_point.empty?
+        filesystem.label = label if label
+        filesystem.uuid = uuid if uuid
+
+        btrfs_setup(filesystem)
```

Comment:
The first line of `btrfs_setup` is already `return unless filesystem.supports_btrfs_subvolumes?` Is there a need to check in both places?